File: packages/sdk/java/core/src/main/java/dev/bugwatch/Fingerprint.java
package dev.bugwatch;

import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.HexFormat;
import java.util.List;

/**
 * Generates a fingerprint (SHA-256 hash) for an error event to group similar errors together.
 */
public final class Fingerprint {

    private Fingerprint() {
    }

    /**
     * Generates a SHA-256 fingerprint for the given error event based on its exception types,
     * messages, and top stack frames.
     *
     * @param event the error event
     * @return hex-encoded SHA-256 hash string
     */
    public static String generate(ErrorEvent event) {
        StringBuilder sb = new StringBuilder();

        ExceptionInfo ex = event.getException();
        if (ex != null) {
            sb.append(ex.type() != null ? ex.type() : "unknown").append(":");
            sb.append(ex.value() != null ? ex.value() : "").append(";");

            if (ex.stacktrace() != null) {
                List<StackFrame> frames = ex.stacktrace();
                // Use the top 5 frames for fingerprinting
                int limit = Math.min(frames.size(), 5);
                for (int i = 0; i < limit; i++) {
                    StackFrame frame = frames.get(i);
                    sb.append(frame.function() != null ? frame.function() : "").append("@");
                    sb.append(frame.lineno()).append(";");
                }
            }
        }

        try {
            MessageDigest digest = MessageDigest.getInstance("SHA-256");
            byte[] hash = digest.digest(sb.toString().getBytes(StandardCharsets.UTF_8));
            return HexFormat.of().formatHex(hash);
        } catch (NoSuchAlgorithmException e) {
            // SHA-256 is guaranteed to be available in all Java implementations
            throw new RuntimeException("SHA-256 not available", e);
        }
    }
}

File: packages/sdk/ruby/lib/bugwatch/fingerprint.rb
# frozen_string_literal: true

require "digest"

module Bugwatch
  module Fingerprint
    def self.generate(event)
      parts = []

      if event.exception
        parts << event.exception.type.to_s
        parts << event.exception.value.to_s

        first_in_app = event.exception.stacktrace.find(&:in_app)
        if first_in_app
          parts << first_in_app.filename.to_s
          parts << first_in_app.lineno.to_s
          parts << first_in_app.function.to_s
        end
      elsif event.message
        parts << event.message
      end

      Digest::SHA256.hexdigest(parts.join("|"))
    end
  end
end
